### DISCUSSION TITLE
Is it good to set action type condition inside a custom redux middleware?

### DISCUSSION THREAD

**Discussion by slamdunkYYY** [reduxjs/redux#4291] [General]
I'm going to create a custom redux. And based on the different scenarios, I'll do some extra inside the middleware. 

I have two options to detect check the condition. Option 1 is checking the action type and based on the action type, to do the extra action. 

```
const customMiddleware= store => next => action => {
    const { type } = action
    const extraActionList = [ "TYPE1", "TYPE2", "TYPE3", "TYPE4", ..., ] // adding all extra action here
    if (extraActionList.includes(type)) {
    .... something extra happen here
    }
    return next(action)
}
````

Another option is instead of set condition inside middleware, I will do something on the action, I add some extra property to the original action.
So the custom middleware will be:
```
const customMiddleware= store => next => action => {
    const { extraAction } = action
    if (extraAction) {
    .... something extra happen here
    }
    return next(action)
}
```
for actions, if the original action object is `{ type: 'TYPE1', payload: 'abc' }`, with those I want to do something extra, I will modify it to `{ type: 'TYPE1', payload: 'abc', extraAction: true }`. But for those not with extra, I will leave them as the old, like `{ type: 'TYPE2', payload: 'abcd' }` without the extra property. 

I want to know which way is better.

**Comment by markerikson** [comment]
I'm not quite sure what the exact question is here.

However, I'd recommend taking a look at our new "listener middleware", which looks like it can do the kinds of things you're asking about:

https://github.com/reduxjs/redux-toolkit/discussions/1648

**Reply by slamdunkYYY** [reply]
This is fantastic! I planed to write a custom middleware to response to a specific action being dispatched. But the action.type condition is quite long as the those specific actions list is long. So I wonder if there is some alternative way for this.

I think this is what I want! Wait for this on the release version soon.

**Reply by markerikson** [reply]
FWIW you can go ahead and try it out right now - `npm i @rtk-incubator/action-listener-middleware`

If you want to handle many different actions the same way, you can use [RTK's "matching" functions](https://redux-toolkit.js.org/api/matching-utilities):

```ts

const matchSeveralActions = isAnyOf(action1, action2, action3);

middleware.startListening({
  matcher: matchSeveralActions,
  effect: (action, listenerApi) => {
    // do stuff here
  }
})
```
